In [1]:
!pip install geocoder
!pip install folium

In [2]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


In [3]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text

In [4]:
soup = BeautifulSoup(source, 'xml')

In [5]:
table=soup.find('table')

In [6]:
from IPython.display import display_html

In [7]:
new_table = str(soup.table)

## Data Cleaning

In [8]:
column_names = ['PostalCode','Borough','Neighborhood']
df = pd.DataFrame(columns = column_names)

In [9]:
df = pd.read_html(new_table)[0]

In [10]:
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

In [11]:
column_names = ['PostalCode','Borough','Neighborhood']
#df = pd.DataFrame(columns = column_names)
df.columns = column_names
# df

In [12]:
# Dropping the rows where Borough is 'Not assigned'
df1 = df[df.Borough != 'Not assigned']

# Combining the neighbourhoods with same Postalcode
df2 = df1.groupby(['PostalCode','Borough'], sort=False).agg(', '.join)
df2.reset_index(inplace=True)

# Replacing the name of the neighbourhoods which are 'Not assigned' with names of Borough
df2['Neighborhood'] = np.where(df2['Neighborhood'] == 'Not assigned',df2['Borough'], df2['Neighborhood'])

df2

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence 

In [17]:
dff = df2.groupby(['PostalCode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
dff.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [18]:
dff.describe()

PostalCode     Borough Neighborhood
count         103         103          103
unique        103          10           99
top           M5S  North York    Downsview
freq            1          24            4

In [19]:
print(dff.shape)
dff.head()

(103, 3)


PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

 ## Import file containing the geospatial data

In [21]:
geo_coords = pd.read_csv('https://cocl.us/Geospatial_data')
geo_coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [22]:
geo_coords.rename(columns = {'Postal Code': 'PostalCode'}, inplace = True)
geo_coords.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [25]:
geo_spa = pd.merge(df2, geo_coords, on = 'PostalCode')

In [26]:
geo_spa  

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                     

In [29]:
Tordf = geo_spa[geo_spa['Borough'].str.contains('North York',regex=False)]
Tordf

PostalCode     Borough                                     Neighborhood  \
0         M3A  North York                                        Parkwoods   
1         M4A  North York                                 Victoria Village   
3         M6A  North York                 Lawrence Manor, Lawrence Heights   
7         M3B  North York                                        Don Mills   
10        M6B  North York                                        Glencairn   
13        M3C  North York                                        Don Mills   
27        M2H  North York                                Hillcrest Village   
28        M3H  North York  Bathurst Manor, Wilson Heights, Downsview North   
33        M2J  North York                     Fairview, Henry Farm, Oriole   
34        M3J  North York                  Northwood Park, York University   
39        M2K  North York                                  Bayview Village   
40        M3K  North York                                        Downsview   
45        M2L  North York                         York Mills, Silver Hills   
46        M3L  North York                                        Downsview   
49        M6L  North York         North Park, Maple Leaf Park, Upwood Park   
50        M9L  North York                                    Humber Summit   
52        M2M  North York                          Willowdale, Newtonbrook   
53        M3M  North York                                        Downsview   
55        M5M  North York                Bedford Park, Lawrence Manor East   
57        M9M  North York                                 Humberlea, Emery   
59        M2N  North York                      Willowdale, Willowdale East   
60        M3N  North York                                        Downsview   
66        M2P  North York                                  York Mills West   
72        M2R  North York                      Willowdale, Willowdale West   

     Latitude  Longitude  
0   43.753259 -79.329656  
1   43.725882 -79.315572  
3   43.718518 -79.464763  
7   43.745906 -79.352188  
10  43.709577 -79.445073  
13  43.725900 -79.340923  
27  43.803762 -79.363452  
28  43.754328 -79.442259  
33  43.778517 -79.346556  
34  43.767980 -79.487262  
39  43.786947 -79.385975  
40  43.737473 -79.464763  
45  43.757490 -79.374714  
46  43.739015 -79.506944  
49  43.713756 -79.490074  
50  43.756303 -79.565963  
52  43.789053 -79.408493  
53  43.728496 -79.495697  
55  43.733283 -79.419750  
57  43.724766 -79.532242  
59  43.770120 -79.408493  
60  43.761631 -79.520999  
66  43.752758 -79.400049  
72  43.782736 -79.442259

# Visualization of the Neighborhoods

In [34]:
import folium
map_NYork = folium.Map(location=[43.651070,-79.347015],zoom_start=10)

for lat,lng,borough,neighborhood in zip(Tordf['Latitude'],Tordf['Longitude'],Tordf['Borough'],Tordf['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_NYork)
map_NYork

In [38]:
address = 'North York,Toronto'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude_x = location.latitude
longitude_y = location.longitude
print('The Geograpical Co-ordinate of York, Toronto are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of York, Toronto are 43.7543263, -79.44911696639593.


### Define Foursquare Credentials and Version

In [39]:
CLIENT_ID = 'W0ZX3JB1GDM42YB1RDEWCTWL4IJFQ45EWAHE1L4D2FSI2PSB' # your Foursquare ID
CLIENT_SECRET = 'YRYFGOCGGD2OW1FJKPMKZIYWEABITQYWRVEDY3XQTK31I4MR' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: W0ZX3JB1GDM42YB1RDEWCTWL4IJFQ45EWAHE1L4D2FSI2PSB
CLIENT_SECRET:YRYFGOCGGD2OW1FJKPMKZIYWEABITQYWRVEDY3XQTK31I4MR


In [40]:
radius = 700 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude_x, 
   longitude_y, 
    radius, 
   LIMIT)
results = requests.get(url).json()

In [41]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['referralId', 'reasons.count', 'reasons.items', 'venue.id',
       'venue.name', 'venue.location.address', 'venue.location.crossStreet',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.labeledLatLngs', 'venue.location.distance',
       'venue.location.postalCode', 'venue.location.cc',
       'venue.location.neighborhood', 'venue.location.city',
       'venue.location.state', 'venue.location.country',
       'venue.location.formattedAddress', 'venue.categories',
       'venue.photos.count', 'venue.photos.groups'],
      dtype='object')

In [42]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

## Nearby Location/Venues

In [43]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name                                   venue.categories  \
0               Grill Gate  [{'id': '4bf58dd8d48988d1c0941735', 'name': 'M...   
1            Wolfie's Deli  [{'id': '4bf58dd8d48988d146941735', 'name': 'D...   
2  Orly Restaurant & Grill  [{'id': '4bf58dd8d48988d115941735', 'name': 'M...   
3              Tim Hortons  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...   
4               Bagel Plus  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...   

   venue.location.lat  venue.location.lng  
0           43.753123          -79.451690  
1           43.754875          -79.442438  
2           43.754493          -79.443507  
3           43.754767          -79.443250  
4           43.755395          -79.440686

## Categories of Nearby Locations/Venues

In [44]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name                 categories        lat        lng
0               Grill Gate   Mediterranean Restaurant  43.753123 -79.451690
1            Wolfie's Deli              Deli / Bodega  43.754875 -79.442438
2  Orly Restaurant & Grill  Middle Eastern Restaurant  43.754493 -79.443507
3              Tim Hortons                Coffee Shop  43.754767 -79.443250
4               Bagel Plus                 Restaurant  43.755395 -79.440686

## Top Ten Categories

In [46]:
# Top 10 Categories
a=pd.Series(nearby_venues.categories)
a.value_counts()[:10]

Pizza Place            2
Convenience Store      1
Deli / Bodega          1
American Restaurant    1
Sushi Restaurant       1
Bus Line               1
Grocery Store          1
Restaurant             1
Fried Chicken Joint    1
Men's Store            1
Name: categories, dtype: int64

In [47]:
 def getNearbyVenues(names, latitudes, longitudes, radius=700):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [49]:
# Nearby Venues
NorthYork_venues = getNearbyVenues(names = Tordf['Neighborhood'],
                                   latitudes = Tordf['Latitude'],
                                   longitudes = Tordf['Longitude']
                                  )

Parkwoods
Victoria Village
Lawrence Manor, Lawrence Heights
Don Mills
Glencairn
Don Mills
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Fairview, Henry Farm, Oriole
Northwood Park, York University
Bayview Village
Downsview
York Mills, Silver Hills
Downsview
North Park, Maple Leaf Park, Upwood Park
Humber Summit
Willowdale, Newtonbrook
Downsview
Bedford Park, Lawrence Manor East
Humberlea, Emery
Willowdale, Willowdale East
Downsview
York Mills West
Willowdale, Willowdale West


In [51]:
print('There are {} Uniques Categories.'.format(len(NorthYork_venues['Venue Category'].unique())))
NorthYork_venues.groupby('Neighborhood').count().head()

There are 118 Uniques Categories.


Neighborhood Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North                     26   
Bayview Village                                                      8   
Bedford Park, Lawrence Manor East                                   28   
Don Mills                                                           29   
Downsview                                                           34   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Bathurst Manor, Wilson Heights, Downsview North                      26   
Bayview Village                                                       8   
Bedford Park, Lawrence Manor East                                    28   
Don Mills                                                            29   
Downsview                                                            34   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Bathurst Manor, Wilson Heights, Downsview North     26              26   
Bayview Village                                      8               8   
Bedford Park, Lawrence Manor East                   28              28   
Don Mills                                           29              29   
Downsview                                           34              34   

                                                 Venue Longitude  \
Neighborhood                                                       
Bathurst Manor, Wilson Heights, Downsview North               26   
Bayview Village                                                8   
Bedford Park, Lawrence Manor East                             28   
Don Mills                                                     29   
Downsview                                                     34   

                                                 Venue Category  
Neighborhood                                                     
Bathurst Manor, Wilson Heights, Downsview North              26  
Bayview Village                                               8  
Bedford Park, Lawrence Manor East                            28  
Don Mills                                                    29  
Downsview                                                    34

## Hot Encoding Features

In [52]:
# one hot encoding
NorthYork_onehot = pd.get_dummies(NorthYork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
NorthYork_onehot['Neighborhood'] = NorthYork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [NorthYork_onehot.columns[-1]] + list(NorthYork_onehot.columns[:-1])
NorthYork_onehot = NorthYork_onehot[fixed_columns]
NorthYork_grouped =  NorthYork_onehot.groupby('Neighborhood').mean().reset_index()
NorthYork_onehot.head(5)

Neighborhood  Accessories Store  Airport  American Restaurant  Art Gallery  \
0    Parkwoods                  0        0                    0            0   
1    Parkwoods                  0        0                    0            0   
2    Parkwoods                  0        0                    0            0   
3    Parkwoods                  0        0                    0            0   
4    Parkwoods                  0        0                    0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Bagel Shop  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Bakery  Bank  Bar  Baseball Field  Beer Store  Bike Shop  Boutique  \
0       0     0    0               0           0          0         0   
1       0     0    0               0           0          0         0   
2       0     0    0               0           0          0         0   
3       0     0    0               0           0          0         0   
4       0     0    0               0           0          0         0   

   Bowling Alley  Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  \
0              0            0                0             0              0   
1              0            0                0             0              0   
2              0            0                0             0              0   
3              0            0                0             0              0   
4              0            0                0             0              0   

   Bus Line  Bus Station  Business Service  Butcher  Café  \
0         0            0                 0        0     0   
1         0            0                 0        0     0   
2         0            0                 0        0     0   
3         0            0                 0        0     0   
4         0            0                 0        0     0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0            0                   0               0   

   Clothing Store  Coffee Shop  Comfort Food Restaurant  Community Center  \
0               0            0                        0                 0   
1               0            0                        0                 0   
2               0            0                        0                 0   
3               0            0                        0                 0   
4               0            0                        0                 0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                           0                  0               0   
3                           1                  0               0   
4                           0                  0               0   

   Deli / Bodega  Department Store  Dessert Shop  Dim Sum Restaurant  Diner  \
0              0                 0             0                   0      0   
1              0                 0             0                   0      0   
2              0                 0             0                   0      0   
3              0                 0             0                   0      0   
4              0                 0             0           

In [53]:
num_top_venues = 5
for hood in NorthYork_grouped['Neighborhood']:
    print("---- "+hood+" ----")
    temp = NorthYork_grouped[NorthYork_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Bathurst Manor, Wilson Heights, Downsview North ----
         venue  freq
0         Park  0.08
1  Coffee Shop  0.08
2         Bank  0.08
3  Bridal Shop  0.04
4     Pharmacy  0.04


---- Bayview Village ----
                venue  freq
0                Bank  0.25
1       Grocery Store  0.12
2        Skating Rink  0.12
3        Intersection  0.12
4  Chinese Restaurant  0.12


---- Bedford Park, Lawrence Manor East ----
                venue  freq
0  Italian Restaurant  0.11
1         Coffee Shop  0.11
2      Sandwich Place  0.07
3                 Pub  0.04
4             Butcher  0.04


---- Don Mills ----
                 venue  freq
0  Japanese Restaurant  0.14
1                  Gym  0.10
2          Coffee Shop  0.07
3           Restaurant  0.07
4           Beer Store  0.07


---- Downsview ----
                   venue  freq
0  Vietnamese Restaurant  0.09
1          Grocery Store  0.09
2            Coffee Shop  0.06
3            Pizza Place  0.06
4            Gas Station  0.06




In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

## Most Common Locations/Venues near Neighborhood

In [56]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = NorthYork_grouped['Neighborhood']

for ind in np.arange(NorthYork_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(NorthYork_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  \
0  Bathurst Manor, Wilson Heights, Downsview North                   Park   
1                                  Bayview Village                   Bank   
2                Bedford Park, Lawrence Manor East     Italian Restaurant   
3                                        Don Mills    Japanese Restaurant   
4                                        Downsview  Vietnamese Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank           Coffee Shop         Grocery Store   
1         Grocery Store                  Café          Skating Rink   
2           Coffee Shop        Sandwich Place         Grocery Store   
3                   Gym            Restaurant           Coffee Shop   
4         Grocery Store           Pizza Place                  Park   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant           Pizza Place              Pharmacy   
1   Japanese Restaurant          Intersection    Chinese Restaurant   
2      Greek Restaurant        Ice Cream Shop     Indian Restaurant   
3            Beer Store             Bike Shop        Clothing Store   
4           Gas Station           Coffee Shop        Sandwich Place   

  8th Most Common Venue      9th Most Common Venue   10th Most Common Venue  
0    Chinese Restaurant  Middle Eastern Restaurant         Community Center  
1     Convenience Store             Cosmetics Shop            Deli / Bodega  
2             Juice Bar               Liquor Store  Comfort Food Restaurant  
3                Office         Chinese Restaurant     Caribbean Restaurant  
4   Fried Chicken Joint                   Pharmacy     Outdoor Supply Store

### K- Means Clustering

In [85]:
# k=5
#toronto_clustering = Tordf.drop(['PostalCode','Borough','Neighborhood'],1)
 
#kmeans = KMeans(n_clusters = k,random_state=0).fit(toronto_clustering)
# kmeans.labels_
# Tordf.insert(0, 'Cluster Labels', kmeans.labels_)

In [86]:
#Tordf
# NorthYork_grouped

In [87]:
# Using K-Means to cluster neighborhood into 3 clusters
k = 3
NorthYork_grouped_clustering = NorthYork_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=k, random_state=0).fit(NorthYork_grouped_clustering)
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 1, 0, 1, 0, 0, 1])

In [102]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

NorthYork_merged = dff.iloc[:8,:]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
NorthYork_merged = NorthYork_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

NorthYork_merged.head()# check the last columns!

ValueError: cannot insert Cluster Labels, already exists

In [103]:
NorthYork_merged

Cluster Labels PostalCode     Borough  \
0                4        M3A  North York   
1                4        M4A  North York   
3                0        M6A  North York   
7                4        M3B  North York   
10               0        M6B  North York   
13               4        M3C  North York   
27               1        M2H  North York   
28               2        M3H  North York   

                                       Neighborhood   Latitude  Longitude  
0                                         Parkwoods  43.753259 -79.329656  
1                                  Victoria Village  43.725882 -79.315572  
3                  Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
7                                         Don Mills  43.745906 -79.352188  
10                                        Glencairn  43.709577 -79.445073  
13                                        Don Mills  43.725900 -79.340923  
27                                Hillcrest Village  43.803762 -79.363452  
28  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259

## Map Clusters

In [104]:
kclusters = 10

In [105]:


# create map
map_clusters = folium.Map(location=[latitude_x, longitude_y], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)
# add markers to the map

markers_colors = []
for lat, lon, nei , cluster in zip(NorthYork_merged['Latitude'], 
                                   NorthYork_merged['Longitude'], 
                                   NorthYork_merged['Neighborhood'], 
                                   NorthYork_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters



['#8000ff', '#4856fb', '#10a2f0', '#2adddd', '#62fbc4', '#9cfba4', '#d4dd80', '#ffa256', '#ff562c', '#ff0000']


In [106]:
df1=NorthYork_merged.loc[NorthYork_merged['Cluster Labels'] == 0,NorthYork_merged.columns[[2] + list(range(5, NorthYork_merged.shape[1]))]]
df2=NorthYork_merged.loc[NorthYork_merged['Cluster Labels'] == 1,NorthYork_merged.columns[[2] + list(range(5, NorthYork_merged.shape[1]))]]
df3=NorthYork_merged.loc[NorthYork_merged['Cluster Labels'] == 2,NorthYork_merged.columns[[2] + list(range(5, NorthYork_merged.shape[1]))]]

In [107]:
NorthYork_Avg_HousingPrice.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'NorthYork_Avg_HousingPrice' is not defined

In [97]:
NorthYork_Avg_HousingPrice.plot(kind='bar',figsize=(24,18),alpha=0.75)

NameError: name 'NorthYork_Avg_HousingPrice' is not defined

## School Rating by Clusters

In [98]:
clusters=pd.DataFrame({"Cluster1":df1["Neighborhood"],
                      "Cluster2":df2["Neighborhood"],
                      "Cluster4":df3["Neighborhood"]})
clusters = clusters.replace(np.nan, '', regex=True)

KeyError: 'Neighborhood'

In [99]:
new_NorthYork=NorthYork_merged.set_index("Neighborhood",drop=True)
#Source:https://www.greatschools.org
NorthYork_school_ratings=pd.DataFrame({"Neighborhood":df["Neighborhood"],
                                      "Top School Rating":[7,9,5,8,10,10,7,10,1,2,1,2,7,2,3,2,6,
                                                           5,4,8,9,9,6,6,4,5,4,6,8,10,8,9,6,2,
                                                           10,2,5,8,9,6,6,10,8,9,1,2,3,4,5,6,9,
                                                           8,5,9,6,9,6,4,8,10,2,5,6,3,9,8,7,
                                                           7,8,5,8,9,1,5,4,7,2,3,6,6,9,4,8,7,
                                                           4,8,9,2,6,4,7,5,10,4,6,8,9,7,5,6,5,8,7
                                                           ]})

ValueError: array length 103 does not match index length 180

In [100]:
Scarborough_school_ratings.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_school_ratings' is not defined

In [101]:
Scarborough_school_ratings.plot(kind='bar',figsize=(16,10),color='green',alpha=0.75);

NameError: name 'Scarborough_school_ratings' is not defined

Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made.